### PR TITLE
More precise errors

### PR DESCRIPTION
**Pull request by Chris-Sigopt**
Move a bunch (but not all) of our generic BadParamErrors to different, more descriptive, error types. 

Generally left some BadParamErrors if the error didn't feel like a `validation` error to me.

### CODE REVIEW

**Review by dwanderson-intel — APPROVED** [review]
LGTM so long as we're still correctly raising/returning `RequestError`-types at the end of API calls (so that the HTTP status code gets back to the caller).

**Comment by Chris-Sigopt** [comment]
> LGTM so long as we're still correctly raising/returning `RequestError`-types at the end of API calls (so that the HTTP status code gets back to the caller).

Yeah, we still intercept every exception inheriting from `SigoptValidationError` at the `prod.py` level and turn it into a BadParamError at the very end- that's one reason I'm not super inclined to kill that particular error overall, we still turn everything into that error when sending it out to the public.

**Comment by dwanderson-intel** [comment]
perfect! yeah, BadParam, with its HTTP code, is still an appropriate thing to throw, and now it's only at the right level. awesome!

File: src/python/zigopt/api/request.py [hunk 1]
```diff
@@ -11,24 +11,24 @@
...
-from libsigopt.aux.errors import MissingParamError
+from libsigopt.aux.errors import InvalidKeyError, InvalidValueError, MissingParamError, SigoptValidationError
 
 
 DEFAULT_PAGING_MAX_LIMIT = 1000
 
 
-def validate_api_input_string(input_str: str) -> str:
-  assert isinstance(input_str, str)
+def validate_api_input_string(input_string):
```
Comment (dwanderson-intel) [inline src/python/zigopt/api/request.py:23]: any reason to lose the typing info?

**Pull request merged**